### ISSUE TITLE
Implicit conversions + tuple instances cause combinatorial explosion in Scala 3

### ISSUE THREAD

**Issue by armanbilge** [typelevel/cats#4035]
Continuing from https://github.com/lampepfl/dotty/issues/13838.

**tl;dr**: the implicit conversions used to implement syntax combined with the auto-generated implicit tuple instances creates a combinatorial explosion that causes compile times to grow exponentially. The reason this affects Scala 3 only is due to its corrected behavior wrt implicit search.

It's looking like probably a wontfix/cantfix for dotty, so I think this is something we should try and fix in cats if possible.

One possible solution, based on Odersky's suggestion:
1. Split existing auto-generated sources for tuple instances across 2/3. For Scala 2, they remain as is; for Scala 3, we keep for bincompat but remove the implicit modifier.
2. On Scala 3 only, re-implement these implicits in terms of `:*`.

An alternate solution, based on a suggestion by Nicolas Stucki, would be to instead split syntax implicit conversions across Scala 2/3 and and re-implement using `given Conversion` on Scala 3 instead.

Possibly, implementing both suggestions would be ideal.

**Comment by nicolasstucki** [comment]
Reimplementing the implicits would look somthing like
```scala
implicit def catsKernelOrderForEmptyTuple: Order[EmptyTuple] = ...
implicit def catsKernelOrderForTuple[H, T <: Tuple](implicit H: Order[H], T: Order[T]): Order[H *: T] = ...
```
which changes the combinatorial search into a linear search.


Reimplementing the implicit conversion would look like
```diff
- implicit def catsSyntaxEq[A: Eq](a: A): Foo[A] = ...
+ given catsSyntaxEq[A: Eq]: Conversion[A, Foo[A]] = ...
```

I would suggest both as they can both help with compilation speeds.

**Comment by joroKr21** [comment]
The example is very confusing to me because you use `Foo` both for the syntax and you provide type class instances for it. That doesn't actually exist in the cats codebase.

Also do you maybe have an example of the combinatorial explosion? What implicit searches does Scala 3 go through? Any pointers how to debug this with Scala 3?

**Comment by armanbilge** [comment]
> The example is very confusing to me because you use `Foo` both for the syntax and you provide type class instances for it.

Sorry, can you clarify what you mean by this?

FWIW this happened in a real code base, and I was able to work around it by avoiding in this situation by avoiding the `cats.syntax.all.*` import: https://github.com/armanbilge/schrodinger/commit/6874abf2fc2739760f7db6aef5d5bb56bc9bfe5c

**Comment by joroKr21** [comment]
I mean those two things:
```scala
  implicit def catsSyntaxEq[A: Eq](a: A): Foo[A] = ???
  implicit def fooEq[A: Eq]: Eq[Foo[A]] = ???
```

`catsSyntaxEq` normally returns a value class `EqSyntax` or whatever that has some methods but it doesn't implement any type classes.

**Comment by joroKr21** [comment]
It would be as implementing `Eq[EqOps[A]]` - it doesn't make sense

**Comment by armanbilge** [comment]
Hmm, in the process of minimizing it looks like it got convoluted.

**Comment by armanbilge** [comment]
@joroKr21 what about this minimization?
https://github.com/lampepfl/dotty/issues/13838#issuecomment-958316719

**Comment by joroKr21** [comment]
Yeah that looks more plausible - but where does the combinatorial explosion come from? 🤔 
I don't see it looking at the code.

**Comment by armanbilge** [comment]
```scala
  def liftF[F[_], A](fa: F[A]): F[Foo[A]] = map(fa)(???)
  def map[F[_], A](ffa: F[Foo[A]])(f: A): Nothing = ???
```

I think because `fa` doesn't match the signature of `map`, it starts looking for implicit conversions, finds the one with the `Eq` constraint, then starts trying to expand the `Order` instances?

**Comment by joroKr21** [comment]
But the implicit conversion gives you an incompatible type - why expand anything?

**Comment by armanbilge** [comment]
/shrug not sure. Using `given Conversion` doesn't have that problem, so it's a problem with implicit conversions?

**Comment by joroKr21** [comment]
Sounds like a problem with the Scala 3 compiler - I think it should be fixed (I think we will still have `implicit def` for a while until libraries switch to Scala 3 only support)

**Comment by armanbilge** [comment]
Yes, just saw your comment there! That sounds great, thanks for your attention to this 😁

**Comment by joroKr21** [comment]
> Yes, just saw your comment there! That sounds great, thanks for your attention to this 😁

Sorry I got confused because Scala 2 doesn't have opaque types - the issue is with the type alias

**Comment by joroKr21** [comment]
Unfortunately I don't think changing the code for Scala 3 is desirable:
```scala
implicit def catsKernelOrderForEmptyTuple: Order[EmptyTuple] = ...
implicit def catsKernelOrderForTuple[H, T <: Tuple](implicit H: Order[H], T: Order[T]): Order[H *: T] = ...
```

because this will be quite inefficient at runtime. Might as well just go for full type class derivation which we are experimenting with in Kittens but there are still issues for example `Mirror` for tuples doesn't work yet.

**Comment by nicolasstucki** [comment]
`catsKernelOrderForTuple` could be a macro that would generate calls to the equivalent `def catsKernelOrderForTupleN`.

**Comment by armanbilge** [comment]
Is this issue still relevant after the fix described in https://github.com/lampepfl/dotty/issues/13838#issuecomment-964546201?

**Comment by joroKr21** [comment]
Martin made this comment:
> I found a way to make the original test cases work without the search limit, by improving the way we recognize underspecified queries. The search limit is still there since there are other test cases where we need it.

But I didn't see what it actually refers to 🤔

**Comment by nicolasstucki** [comment]
>> I found a way to make the original test cases work without the search limit, by improving the way we recognize underspecified queries. The search limit is still there since there are other test cases where we need it.
>
> But I didn't see what it actually refers to 🤔

It was the minimized version that can be found in
https://github.com/dotty-staging/dotty/blob/739dec3c774c372408798b364574e8096db0735b/tests/neg/i13838a.scala

The fix was quite specific to this use case. The combinatorial explosion could still happen in theory, but I'm not sure how to trigger it. At least we also added extra tools in that same PR to be able to detect those explosions avoid looping infinitely. This means that if someone hits this issue again it will be simpler to diagnose.

**Comment by joroKr21** [comment]
I need to take a closer look at the solution to understand what you mean. IMO it should be very general - the problem is that we were attempting an implicit search for an erroneous tree that we already knew doesn't type check.

**Comment by joroKr21** [comment]
Yeah I think Martin's solution is quite general. The code which requires the new `-Ximplicit-search-limit` doesn't exist in Cats and doesn't make sense in general.

**Comment by armanbilge** [comment]
> The code which requires the new -Ximplicit-search-limit doesn't exist in Cats and doesn't make sense in general.

Sounds like we can close then :) thanks everyone!
